File: app/controllers/question/ExamMaterialController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.question;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import controllers.base.SectionQuestionHandler;
import io.ebean.DB;
import io.ebean.text.PathProperties;
import java.util.Optional;
import java.util.Set;
import models.sections.ExamMaterial;
import models.sections.ExamSection;
import models.user.User;
import org.springframework.beans.BeanUtils;
import play.mvc.Http;
import play.mvc.Result;
import security.Authenticated;
import validation.core.Attrs;

public class ExamMaterialController extends QuestionController implements SectionQuestionHandler {

    private ExamMaterial parseFromBody(JsonNode body) {
        ExamMaterial em = new ExamMaterial();
        em.setName(body.get("name").asText());
        em.setAuthor(body.path("author").asText(null));
        em.setIsbn(body.path("isbn").asText(null));
        return em;
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result createMaterial(Http.Request request) {
        ExamMaterial em = parseFromBody(request.body().asJson());
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        em.setCreatorWithDate(user);
        em.setModifierWithDate(user);
        em.save();
        return ok(em);
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result listMaterials(Http.Request request) {
        Set<ExamMaterial> materials = DB.find(ExamMaterial.class)
            .where()
            .eq("creator", request.attrs().get(Attrs.AUTHENTICATED_USER))
            .findSet();
        return ok(materials, PathProperties.parse("(*)"));
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result removeMaterial(Long materialId, Http.Request request) {
        ExamMaterial em = DB.find(ExamMaterial.class, materialId);
        if (em == null || !em.getCreator().equals(request.attrs().get(Attrs.AUTHENTICATED_USER))) {
            return notFound();
        }
        DB.delete(ExamMaterial.class, materialId);
        return ok();
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result updateMaterial(Long materialId, Http.Request request) {
        ExamMaterial dst = DB.find(ExamMaterial.class, materialId);
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (dst == null || !dst.getCreator().equals(user)) {
            return notFound();
        }
        ExamMaterial src = parseFromBody(request.body().asJson());
        BeanUtils.copyProperties(src, dst, "id", "examSections", "objectVersion");
        dst.update();
        return ok();
    }

    private Optional<ExamSection> getSection(Long sectionId, User user) {
        return DB.find(ExamSection.class).where().idEq(sectionId).eq("exam.examOwners", user).findOneOrEmpty();
    }

    private Optional<Result> getOwnershipError(ExamMaterial em, User user) {
        return em == null || !em.getCreator().equals(user) ? Optional.of(notFound()) : Optional.empty();
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result addMaterialForSection(Long sectionId, Long materialId, Http.Request request) {
        ExamMaterial em = DB.find(ExamMaterial.class, materialId);
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        return getOwnershipError(em, user).orElseGet(() -> {
            Optional<ExamSection> oes = getSection(sectionId, user);
            if (oes.isPresent()) {
                ExamSection es = oes.get();
                es.getExamMaterials().add(em);
                es.update();
                return ok();
            }
            return notFound();
        });
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result removeMaterialFromSection(Long sectionId, Long materialId, Http.Request request) {
        ExamMaterial em = DB.find(ExamMaterial.class, materialId);
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        return getOwnershipError(em, user).orElseGet(() -> {
            Optional<ExamSection> oes = getSection(sectionId, user);
            if (oes.isPresent()) {
                ExamSection es = oes.get();
                es.getExamMaterials().remove(em);
                es.update();
                return ok();
            }
            return notFound();
        });
    }
}

File: app/controllers/question/ExamMaterialController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.question

import impl.SectionQuestionHandler
import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.sections.{ExamMaterial, ExamSection}
import models.user.{Role, User}
import play.api.Logging
import play.api.libs.json.*
import play.api.mvc.*
import security.scala.Auth
import security.scala.Auth.{AuthenticatedAction, authorized}
import system.AuditedAction

import javax.inject.Inject
import scala.concurrent.ExecutionContext
import scala.jdk.CollectionConverters.*

class ExamMaterialController @Inject() (
    authenticated: AuthenticatedAction,
    audited: AuditedAction,
    val controllerComponents: ControllerComponents
)(implicit ec: ExecutionContext)
    extends BaseController
    with SectionQuestionHandler
    with DbApiHelper
    with JavaApiHelper
    with Logging:

  private def parseFromBody(body: JsValue): ExamMaterial =
    val em = new ExamMaterial()
    em.setName((body \ "name").as[String])
    (body \ "author").asOpt[String].foreach(em.setAuthor)
    (body \ "isbn").asOpt[String].foreach(em.setIsbn)
    em

  def createMaterial(): Action[JsValue] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN)))(parse.json) { request =>
      val em   = parseFromBody(request.body)
      val user = request.attrs(Auth.ATTR_USER)
      em.setCreatorWithDate(user)
      em.setModifierWithDate(user)
      em.save()
      Ok(em.asJson)
    }

  def listMaterials(): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN))) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      val materials = DB
        .find(classOf[ExamMaterial])
        .where()
        .eq("creator", user)
        .distinct
      Ok(materials.asJson(PathProperties.parse("(*)")))
    }

  def removeMaterial(materialId: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN))) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      Option(DB.find(classOf[ExamMaterial], materialId)) match
        case None                                    => NotFound
        case Some(em) if !em.getCreator.equals(user) => NotFound
        case Some(_) =>
          DB.delete(classOf[ExamMaterial], materialId)
          Ok
    }

  def updateMaterial(materialId: Long): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN)))(parse.json) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      Option(DB.find(classOf[ExamMaterial], materialId)) match
        case None                                      => NotFound
        case Some(dst) if !dst.getCreator.equals(user) => NotFound
        case Some(dst) =>
          val src = parseFromBody(request.body)
          // Copy properties manually (Scala alternative to BeanUtils.copyProperties)
          dst.setName(src.getName)
          Option(src.getAuthor).foreach(dst.setAuthor)
          Option(src.getIsbn).foreach(dst.setIsbn)
          dst.update()
          Ok
    }

  private def getSection(sectionId: Long, user: User): Option[ExamSection] =
    DB.find(classOf[ExamSection])
      .where()
      .idEq(sectionId)
      .eq("exam.examOwners", user)
      .find

  private def getOwnershipError(em: ExamMaterial, user: User): Option[Result] =
    if em == null || !em.getCreator.equals(user) then Some(NotFound)
    else None

  def addMaterialForSection(sectionId: Long, materialId: Long): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN))) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      Option(DB.find(classOf[ExamMaterial], materialId)) match
        case None => NotFound
        case Some(em) =>
          getOwnershipError(em, user) match
            case Some(error) => error
            case None =>
              getSection(sectionId, user) match
                case None => NotFound
                case Some(es) =>
                  es.getExamMaterials.add(em)
                  es.update()
                  Ok
    }

  def removeMaterialFromSection(sectionId: Long, materialId: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN))) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      Option(DB.find(classOf[ExamMaterial], materialId)) match
        case None => NotFound
        case Some(em) =>
          getOwnershipError(em, user) match
            case Some(error) => error
            case None =>
              getSection(sectionId, user) match
                case None => NotFound
                case Some(es) =>
                  es.getExamMaterials.remove(em)
                  es.update()
                  Ok
    }
